1.17 Cleanup Tag Loading

Cyborgmas opened the pull request:
There was some refactors of how tags were loaded, with LoaderInfo being a wrapper for a CompletableFuture, so instead of making the custom CF, the patch is simplified to create a LoaderInfo instead.

Cyborgmas committed:
cleanup tag loading

SizableShrimp commented on src/main/resources/META-INF/accesstransformer.cfg:
```diff
@@ -398,4 +398,6 @@ public net.minecraft.world.level.newbiome.layer.Layers m_76729_(JLnet/minecraft/
\ No newline at end of file
+private-f net.minecraft.world.level.storage.loot.LootPool f_79029_ # bonusRolls
+public net.minecraft.tags.TagManager$LoaderInfo
+public net.minecraft.tags.TagManager$LoaderInfo <init>(Lnet/minecraft/tags/StaticTagHelper;Ljava/util/concurrent/CompletableFuture;)V
```
These should be alphabetized to the right spot in the file with the empty line removed, then run checkAts.

Cyborgmas committed:
checkats

pupnewfster commented on patches/minecraft/net/minecraft/tags/TagManager.java.patch:
```diff
@@ -39,3 +17,11 @@
+          }, p_144577_);
+          return new TagManager.LoaderInfo<>(p_144578_, completablefuture);
+       } else {
++         if (true) return net.minecraftforge.common.ForgeTagHandler.makeCustomTagTypeLoader(p_144578_, p_144576_, p_144577_, customTagTypeReaders);
```
This seems to be silencing a vanilla warning. While I agree if we are able to make a custom tag type loader we shouldn't warn that it isn't found, but we probably should still have a warning if it failed (be it that it couldn't find the registry or that the registry doesn't support custom tag types).

pupnewfster commented on src/main/java/net/minecraftforge/common/ForgeTagHandler.java:
```diff
@@ -289,21 +282,6 @@ public static void populateTagCollectionManager()
-        {
-            builder.put(info.tagType, info.reader.build(info.tagBuilders));
-        }
-        customTagTypes = builder.build();
```
I haven't looked in great detail yet how tags work/changed in 1.17 but it seems like `ForgeTagHandler.customTagTypes` may not being updated anymore post reload due to the removal of this line?

Cyborgmas commented:
So small issue, I am adding the custom tags to the vanilla construct (TagContainer) so it gets serialized/deserialized by vanilla, i can patch in deserialization, but it wont work for vanilla clients however it only logs an error, it doesnt crash (at least up to the point i checked).
if the vanilla error logging is fine, ill try and continue with this, because I think it will reduce the overall code needed for custom tags to work.

pupnewfster commented:
For vanilla clients we could always use the to vanilla packet filter forge has to remove the "invalid" types from it. But if vanilla's tag changes make a cleaner/easier solution to custom tag types possible then it is probably worth doing so.

Cyborgmas committed:
took out some parts, now to test what breaks

Cyborgmas committed:
it didnt break!

Cyborgmas committed:
better info

Cyborgmas committed:
forgot patch

Cyborgmas commented:
Vanilla stopped hardcoding to 4 types of tags, allowing for a much easier handling of custom tag types. 
It's also completely vanilla compatible, the error i mentioned will only happen for non-vanilla registries.

pupnewfster commented on patches/minecraft/net/minecraft/client/multiplayer/ClientPacketListener.java.patch:
```diff
@@ -77,18 +77,17 @@
 -      Multimap<ResourceKey<? extends Registry<?>>, ResourceLocation> multimap = StaticTags.m_13283_(tagcontainer);
 +      boolean vanillaConnection = net.minecraftforge.fmllegacy.network.NetworkHooks.isVanillaConnection(f_104885_);
-+      Multimap<ResourceKey<? extends Registry<?>>, ResourceLocation> multimap = vanillaConnection ? StaticTags.m_13283_(net.minecraftforge.common.ForgeTagHandler.withNoCustom(tagcontainer)) : StaticTags.validateVanillaTags(tagcontainer);//Forge: If we are connecting to vanilla validate all tags to properly validate custom tags the client may "require", and if we are connecting to forge only vali ...
++      Multimap<ResourceKey<? extends Registry<?>>, ResourceLocation> multimap = StaticTags.validateTags(tagcontainer, vanillaConnection);//Forge: If we are connecting to vanilla, custom tag types will be skipped over in validation as the server can't have them
```
This comment (and I presume behavior based on what the comment is), is not actually quite right. The idea is that even for custom tags the connection is terminated if the tags are declared as being required. For custom tag types that the client doesn't necessarily require to be present the idea was to use forge's optional named tag system the same as how you can declare vanilla tag types on clients in a way that won't crash when connecting to a vanilla server/server without the mod.

pupnewfster commented on patches/minecraft/net/minecraft/server/ServerResources.java.patch:
```diff
@@ -12,7 +12,7 @@
     public void m_136179_() {
        this.f_136148_.m_13480_().m_13431_();
-+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.TagsUpdatedEvent.CustomTagTypes(f_136148_.m_13480_()));
++      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.TagsUpdatedEvent(f_136148_.m_13480_()));
```
This line/patch can just be removed as the above line fires the event in `bindToGlobal`

pupnewfster commented on patches/minecraft/net/minecraft/tags/StaticTags.java.patch:
```diff
@@ -21,42 +21,30 @@
 +   }
 +
-+   public static Multimap<ResourceKey<? extends Registry<?>>, ResourceLocation> validateVanillaTags(TagContainer tagCollectionSupplier) {
++   public static Multimap<ResourceKey<? extends Registry<?>>, ResourceLocation> validateTags(TagContainer tagContainer, boolean isVanilla) {
```
See my comment that I put where this is called, but also maybe this should be called `getAllMissingTags` the same as vanilla's method is (maybe even using an exec for letting the tag container param get mapped)

pupnewfster commented on src/main/java/net/minecraftforge/common/ForgeTagHandler.java:
```diff
@@ -21,54 +21,42 @@
-import java.util.Optional;
-import java.util.Set;
+
+import java.util.*;
```
Forge tries to avoid wildcard imports (there are a couple other ones being added into this class as well)

pupnewfster commented on patches/minecraft/net/minecraft/tags/TagManager.java.patch:
```diff
@@ -8,34 +8,20 @@
+          return new TagManager.LoaderInfo<>(p_144578_, completablefuture);
+       } else {
++         if (net.minecraftforge.common.ForgeTagHandler.getCustomTagTypeNames().contains(p_144578_.m_144338_().m_135782_()))
++            return net.minecraftforge.common.ForgeTagHandler.makeCustomTagTypeLoader(p_144578_, p_144576_, p_144577_, customTagTypeReaders);
```
Is there a reason this calls a one line helper method that doesn't even query anything native to ForgeTagHandler instead of just inlining the single line here? Doing so would allow for removing the ATs this PR is adding exposing LoaderInfo.

pupnewfster commented on src/main/java/net/minecraftforge/common/ForgeTagHandler.java:
```diff
@@ -246,166 +224,44 @@ public static void setCustomTagTypes(Set<ResourceLocation> customTagTypes)
-
-    /**
-     * Add all the missing optional tags back into the custom tag types tag collections
+     * Creates a map for serialization, copied from {@link TagCollection#serializeToNetwork} but adapted it since it requires
+     * a normal registry.
```
I am not sure which is better/less maintenance, but in theory this could make use of `TagCollection#serializeToNetwork` by making use of the backing vanilla registries all the custom types have and passing it to that. See `ForgeRegistyTagProvider#wrapRegistry`.

pupnewfster commented on src/main/java/net/minecraftforge/common/ForgeTagHandler.java:
```diff
@@ -246,166 +224,44 @@ public static void setCustomTagTypes(Set<ResourceLocation> customTagTypes)
-        }
+            idMap.put(name, idList);
+        });
+        return new TagCollection.NetworkPayload(idMap);
```
Given this method is just used in one place, if we end up going with this way of doing it over using the backing vanilla registry, this should probably just return the idMap so that `TagCollection.NetworkPayload` constructor does not have to be AT'd to being visible/public. (Either way the AT should be able to be removed as if we do switch to the `TagCollection#serializeToNetwork` method then we won't need access to the constructor either)

pupnewfster reviewed (commented):
I haven't done any testing yet, but there are some things that potentially should be addressed. In addition to my comments on specific parts of the code, the following things potentially should be changed as well due to having a similar concept/target as this PR:
- `ForgeRegistyTagProvider` should potentially have fallback handling for the new vanilla tag type of `game_event`. Though that may(?) be slightly outside the scope of this PR due to the fact the GAME_EVENT registry would need to be taken over by forge, which may be better suited to being in its own separate PR that primarily is targetting taking over various new 1.17 registries that forge can easily take over.
- The two `createOptional` methods we patch into all the vanilla tag types that existed in 1.16 should also be patched into the new type vanilla supports (GameEventTags)
- Similarly to how forge patches, `BlockTagsProvider`, `EntityTypeTagsProvider`, `FluidTagsProvider`, and `ItemTagsProvider`, vanilla's new `GameEventTagsProvider` should probably be patched with the extra constructor.

Cyborgmas committed:
address feedback

add some support for game event tags
cleanup custom tag type serialization & validation

Cyborgmas committed:
add packet filter

reduces packet size
prevents rare error logging

Cyborgmas committed:
renamed methods inline with vanilla: create -> bind

Cyborgmas committed:
Merge remote-tracking branch 'origin/1.17.x' into tag-cleanup

Cyborgmas committed:
rename create to bind

Cyborgmas committed:
Merge remote-tracking branch 'origin/1.17.x' into tag-cleanup

Cyborgmas committed:
remove at entries, no longer used

Cyborgmas commented:
So did not add GAME_EVENT to forge registries, that will be cleaner in another PR I think.
Renamed a bunch of methods, hopefully it's fine. Not sure about the ones in GameData though, thought it made more sense, but i can revert if needed

pupnewfster commented on patches/minecraft/net/minecraft/data/tags/GameEventTagsProvider.java.patch:
```diff
@@ -0,0 +1,12 @@
+--- a/net/minecraft/data/tags/GameEventTagsProvider.java
++++ b/net/minecraft/data/tags/GameEventTagsProvider.java
+@@ -11,6 +_,9 @@
+    public GameEventTagsProvider(DataGenerator p_176826_) {
```
This constructor should be deprecated

pupnewfster commented on src/main/resources/forge.exc:
```diff
@@ -16,6 +16,7 @@ net/minecraft/data/tags/BlockTagsProvider.<init>(Lnet/minecraft/data/DataGenerat
 net/minecraft/data/tags/EntityTypeTagsProvider.<init>(Lnet/minecraft/data/DataGenerator;Ljava/lang/String;Lnet/minecraftforge/common/data/ExistingFileHelper;)V=|p_126517_,modId,existingFileHelper
 net/minecraft/data/tags/FluidTagsProvider.<init>(Lnet/minecraft/data/DataGenerator;Ljava/lang/String;Lnet/minecraftforge/common/data/ExistingFileHelper;)V=|p_126523_,modId,existingFileHelper
 net/minecraft/data/tags/ItemTagsProvider.<init>(Lnet/minecraft/data/DataGenerator;Lnet/minecraft/data/tags/BlockTagsProvider;Ljava/lang/String;Lnet/minecraftforge/common/data/ExistingFileHelper;)V=|p_126530_,p_126531_,modId,existingFileHelper
+net/minecraft/data/tags/GameEventTagsProvider.<init>(Lnet/minecraft/data/DataGenerator;Ljava/lang/String;Lnet/minecraftforge/common/data/ExistingFileHelper;)V=|p_126517_,modId,existingFileHelper
```
This should be above ItemTagsProvider as G comes between F and I

pupnewfster commented on patches/minecraft/net/minecraft/tags/TagContainer.java.patch:
```diff
@@ -1,22 +1,21 @@
+       this.m_144436_(new TagContainer.CollectionConsumer() {
+          public <T> void m_142574_(ResourceKey<? extends Registry<T>> p_144481_, TagCollection<T> p_144482_) {
+             Optional<? extends Registry<T>> optional = p_144443_.m_6632_(p_144481_);
++            optional = optional.isPresent() ? optional : net.minecraftforge.common.ForgeTagHandler.getWrapperRegistry(p_144481_);
```
I feel like this may be slightly cleaner moving the optional.isPresent() -> optional return into the getWrapperRegistry method

Cyborgmas committed:
minor feedback changes

Cyborgmas committed:
more feedback adressed, more create -> bind renames

Cyborgmas committed:
be smarter and remove cast

Cyborgmas committed:
fix conflict

pupnewfster committed:
- Fix custom tag types of vanilla registry types not loading properly due to not setting the proper directory the tags folder prefix is no longer implicit
- Remove patched in is(Tag) methods following vanilla's lead for Item, as tag.contains(thing) can be used just as easily as thing.is(tag)
- Minor cleanup to custom tag type test mod

pupnewfster committed:
Fix inverted filter check, and add reverse tag wrapper to GameEvent

pupnewfster committed:
Cleanup patches related to tags some, and add missing AT to GameEventTags#bind

Cyborgmas committed:
Merge pull request #1 from pupnewfster/tag-cleanup

Fix some minor issues and perform a bit of cleanup

pupnewfster commented:
I believe this PR is ready to be assigned but as I made some changes myself to fix some issues I ran into I am going to label this as triage for now instead so that another triage member can review it/at least my changes.

A more complete summary of changes in this PR now that it is complete is as follows:
- Implement patches related to tags for GameEvent that already existed for the other 4 vanilla tag types
- Rewrite hooks for custom tags to be more inline with and take advantage of how vanilla does tag types due to vanilla not hard coding tags quite as much
- Fixed bug 1.17 porting bug causing custom tag types at least for forge patched in ones of vanilla registry types to not actually load due to not specifying the `tags/` sub directory when creating the tag type. Vanilla no longer implicitly includes the tags folder
- Breaking: Removed sub even types of TagsUpdatedEvent for Vanilla and Custom as they are all loaded in one place now with the above change instead of being slightly offset in timing
- Breaking: Removed no longer needed packets for syncing custom tag types as it makes use of vanilla's system now
- ~Breaking: Renamed methods for creating optional tags to be more in line with vanilla's naming scheme instead of still being based on mcp's naming scheme~
- ~Breaking: Removed patched in methods for BlockEntityType, Enchantment, and Potion for checking if they are in a tag. This follows the change vanilla made to Item and EntityType as there is not much benefit to having a `x#in(tag)` helper method instead of just directly calling `tag#contains(x)`~
- ~Breaking: Renamed/changed signatures slightly of the methods in GameData for getting the wrapper registries to not have a bunch of unused params (these methods shouldn't be getting called by modders in the first place, but mentioning it as a breaking change regardless)~
- Rerun test datagen as it never got reran after the tag folder for block entities got changed from when it was for tile entities
- Minor cleanup to the custom tag type test mod

Edit: Forgot to include the cases I tested:
- Forge client + Forge server
- Forge client + Vanilla server
- Vanilla client + Forge server

I did not test client/server pre this change with post this change iven the network sync is changing from custom to vanilla I suspect it may behave somewhat oddly and as breaking changes are still allowed (and this PR has a good number), it is not worth struggling with as I am unsure if there even is a way to make sure it is handled properly.

LexManos commented on patches/minecraft/net/minecraft/tags/BlockTags.java.patch:
```diff
@@ -4,12 +4,12 @@
++       return bindOptional(name, null);
 +   }
 +
-+   public static net.minecraftforge.common.Tags.IOptionalNamedTag<Block> createOptional(net.minecraft.resources.ResourceLocation name, @javax.annotation.Nullable java.util.Set<java.util.function.Supplier<Block>> defaults) {
```
Honestly these changes are unneeded
Create, bind, of, could just called it 'optional'
We don't need to break even more things for no real reason.

LexManos commented on src/main/java/net/minecraftforge/registries/GameData.java:
```diff
@@ -199,7 +199,7 @@ public static void init()
     }
 
-    public static <T extends IForgeRegistryEntry<T>> MappedRegistry<T> getWrapper(ResourceKey<? extends Registry<T>> key, Lifecycle lifecycle)
+    public static <T extends IForgeRegistryEntry<T>> MappedRegistry<T> getWrapper(ResourceKey<? extends Registry<T>> key)
```
Honestly, these changes are not needed. Yes they arguments are unused, but these are ease of use internal functions for our patches. Not something modders should be calling. Also, it seems completely unrelated to this PR.

Cyborgmas committed:
revert some name changes

Cyborgmas committed:
revert some more names

Cyborgmas committed:
Merge remote-tracking branch 'origin/1.17.x' into tag-cleanup

Cyborgmas committed:
fix conflict

Cyborgmas committed:
Merge remote-tracking branch 'origin/1.17.x' into tag-cleanup

Cyborgmas committed:
add back `is` methods
